### PR TITLE
refactor: inject curried `runPkgManagerInstall` to package installers

### PR DESCRIPTION
**Pull request by ochicf**
# Inject curried `runPkgManagerInstall` to package installers

- [x] I reviewed linter warnings + errors, resolved formatting, types and other issues related to my work
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

I noticed that every installer had to call `runPkgManagerInstall` with a lot of configuration, but IMHO the only configuration they should specify is the `packages` list. So, this MR adds a curryable version of `runPkgManagerInstall` that is curried by `createProject` and injected to every installer as an option. This injected function has been curried with all of the necessary configuration, allowing to clear the call in each installer.

Benefits:
- cleaner code of each installer
- adding new options to `runPkgManagerInstall` will not force adding that parameter in each installer, but instead just once in the `createProject.ts` helper where it's curried
- curried options serve only as defaults, so each installer could still override the options if needed

Another small change added here passing the entirety of the options received in `installPackages.ts` to each package installer, since they share the same shape. This will also prevent from having to pass newly added options here.

### CODE REVIEW

**Comment by ochicf** [comment]
Not sure about the naming convention with currying functions. I've named `curryRunPkgManagerInstall` because it's the most obvious but maybe you prefer something different?

I've also kept the original `runPkgManagerInstall` export, though it's not used outside of that module anymore. Maybe it causes confusion?

**Comment by nexxeln** [comment]
Oh wow this is much cleaner.

**Review by nexxeln — APPROVED** [review]

**Review by juliusmarminge — APPROVED** [review]
Looks good.

I actually had a draft to remove the `runPackageManagerInstall` entirely and instead add all the packages to package.json and then run a single `npm/pnpm/yarn install` command at the end, but haven't had time to complete it.

File: src/installers/prisma.ts [hunk 1]
```diff
@@ -4,27 +4,19 @@ import path from "path";
...
+  runPkgManagerInstall,
   pkgManager,
   packages,
   noInstall,
 }) => {
   await runPkgManagerInstall({
-    pkgManager,
-    projectDir,
     packages: ["prisma"],
-    devMode: true,
```
Comment (ochicf) [inline src/installers/prisma.ts:19]: I missed this, so this is not correct.

File: src/installers/tailwind.ts [hunk 2]
```diff
@@ -2,19 +2,13 @@ import type { Installer } from "./index.js";
...
   projectDir,
-  pkgManager,
-  noInstall,
+  runPkgManagerInstall,
 }) => {
   await runPkgManagerInstall({
-    pkgManager,
-    projectDir,
     packages: ["tailwindcss", "postcss", "autoprefixer"],
-    devMode: true,
```
Comment (ochicf) [inline src/installers/tailwind.ts:16]: Same as previous comment.

**Comment by ochicf** [comment]
Just noticed in another branch that some installs were actually passing `devMode` and I missed them in the refactor.

File: src/utils/runPkgManagerInstall.ts [hunk 3]
```diff
@@ -50,3 +54,26 @@ export const runPkgManagerInstall = async (opts: {
...
+export type CurryRunPkgManagerInstallOptions = Omit<
+  RunPkgManagerInstallOptions,
+  "packages"
+>;
+
+export type CurriedRunPkgManagerInstallOptions =
+  Partial<CurryRunPkgManagerInstallOptions> &
+    Omit<RunPkgManagerInstallOptions, keyof CurryRunPkgManagerInstallOptions>;
+
+export const curryRunPckgManagerInstall = (
```
Comment (juliusmarminge) [inline src/utils/runPkgManagerInstall.ts:67]:
```suggestion
export const curryRunPkgManagerInstall = (
```

File: src/utils/runPkgManagerInstall.ts [hunk 4]
```diff
@@ -50,3 +54,26 @@ export const runPkgManagerInstall = async (opts: {
...
+  baseOptions: CurryRunPkgManagerInstallOptions,
+) => {
+  const curriedRunPckgManagerInstall = async (
```
Comment (juliusmarminge) [inline src/utils/runPkgManagerInstall.ts:70]:
```suggestion
  const curriedRunPkgManagerInstall = async (
```

File: src/utils/runPkgManagerInstall.ts [hunk 5]
```diff
@@ -50,3 +54,26 @@ export const runPkgManagerInstall = async (opts: {
...
+    options: CurriedRunPkgManagerInstallOptions,
+  ) =>
+    runPkgManagerInstall({
+      ...baseOptions,
+      ...options,
+    });
+
+  return curriedRunPckgManagerInstall;
```
Comment (juliusmarminge) [inline src/utils/runPkgManagerInstall.ts:78]:
```suggestion
  return curriedRunPkgManagerInstall;
```

File: src/helpers/createProject.ts [hunk 6]
```diff
@@ -18,12 +19,30 @@ export const createProject = async ({
 }: CreateProjectOptions) => {
   const pkgManager = getUserPkgManager();
   const projectDir = path.resolve(process.cwd(), projectName);
+  const runPkgManagerInstall = curryRunPckgManagerInstall({
```
Comment (juliusmarminge) [inline src/helpers/createProject.ts:22]:
```suggestion
  const runPkgManagerInstall = curryRunPkgManagerInstall({
```

File: src/helpers/createProject.ts [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import type { PkgInstallerMap } from "../installers/index.js";
 import path from "path";
 import { getUserPkgManager } from "../utils/getUserPkgManager.js";
+import { curryRunPckgManagerInstall } from "../utils/runPkgManagerInstall.js";
```
Comment (juliusmarminge) [inline src/helpers/createProject.ts:4]:
```suggestion
import { curryRunPkgManagerInstall } from "../utils/runPkgManagerInstall.js";
```

**Review by juliusmarminge — CHANGES_REQUESTED** [review]
Found a typo

**Comment by ochicf** [comment]
Oh sorry just after I pushed I saw that there was the commit suggestion option 🤦‍♂️ Never saw that before, will use next time.

**Review by juliusmarminge — APPROVED** [review]

**Pull request merged**